62 z ale remapping

nicjhan opened the pull request:
This pull request will change results. The line that causes the change is: 

 src/ALE/MOM_regridding.F90 122

Without this conservation checks fail when there are lots of 'squashed' layers. In model coords these have thickness 1e-10, whereas the new grid uses 1e-3

adcroft-gfdl commented:
The changes to MOM_regridding lead to NaNs in several of the ALE tests. I think this is why I had set the default minimum to that value (for stability). Can you revert that default value and add a method setMinThickness that you can use from within the diagnostics to set it to the tiny value you need?

adcroft-gfdl commented:
Doh! There is a method already - setRegriddingMinimumThickness() which is how MOM_ALE.F90 sets it for the model at run-time.

nicjhan commented:
OK, thanks I have fixed that. I'm no longer getting any results (timestats) changes from master. I still don't completely understand why remapping conservation checks are failing when using minthickness of 1e-3).

adcroft-gfdl commented on src/core/MOM.F90:
```diff
@@ -1920,6 +1918,10 @@ subroutine initialize_MOM(Time, param_file, dirs, CS, Time_in)
 
+  ! Initialise the diag mediator, this needs to occur _after_ CS%h has been
+  ! allocated.
+  call diag_mediator_init(G, CS%h, param_file, diag)
```
How about providing a method to associate h, instead of moving this call so far?

adcroft-gfdl commented on src/diagnostics/MOM_diag_to_Z.F90:
```diff
@@ -862,11 +862,11 @@ subroutine register_Z_tracer_low(tr_ptr, name, long_name, units, standard_name,
   CS%missing_tr(m) = CS%missing_value ! This could be changed later, if desired.
   if (CS%nk_zspace > 0) then
-    CS%id_tr(m) = register_diag_field('ocean_model_z', name, CS%axesTz, Time,           &
+    CS%id_tr(m) = register_diag_field('ocean_model_old_z', name, CS%axesTz, Time,           &
```
Let's keep the ocean_model_z intact and use a new module name while we're developing this feature. We'll runn everything through the analysis suite before making the switch.

adcroft-gfdl commented on src/framework/MOM_diag_mediator.F90:
```diff
@@ -31,19 +31,33 @@ module MOM_diag_mediator
+
 implicit none ; private
 
+#include <MOM_memory.h>
```
Needed?

adcroft-gfdl commented on src/framework/MOM_diag_mediator.F90:
```diff
@@ -219,6 +244,25 @@ subroutine set_axes_info(G, param_file, diag_cs, set_vertical)
+                 "The number of depth-space levels.  This is determined \n"//&
+                 "from the size of the variable zw in the output grid file.", &
+                 units="nondim")
+  endif
```
Can we introduce a more succinct syntax such used for "ALE_COORDINATE_CONFIG?
e.g.
ocean_model_new_z,FILE:vgrid.nc,zt
and use "call MOM_read_data(trim(fileName), trim(varName), dz)"

nicjhan commented:
Thanks for your comments, I've fixed those things. There's one thing I'm not sure about. I've changed the grid definition config to be like:

DIAG_REMAP_Z_GRID_DEF = "FILE:<file>,<variable>" e.g. 
DIAG_REMAP_Z_GRID_DEF = "FILE:OM3_zgrid.nc,zw"

This means that when we move to other vertical coords there will need to be another config e.g.:

DIAG_REMAP_SIGMA_GRID_DEF = "FILE:OM3_sigmagrid.nc,si"

Also if we want to use a generated grid, can do:
DIAG_REMAP_Z_GRID_DEF = "UNIFORM:<thickness>"

I'm not sure if this is what you had in mind.

adcroft-gfdl commented:
@nicjhan  That's exactly what I had in mind.
